db/post_migrate/20240307180905_migrate_devise_two_factor_secrets.rb

```diff
@@ -0,0 +1,33 @@
+# frozen_string_literal: true
+
+class MigrateDeviseTwoFactorSecrets < ActiveRecord::Migration[7.1]
+  disable_ddl_transaction!
+
+  class MigrationUser < ApplicationRecord
+    self.table_name = :users
+  end
+
+  def up
+    users_with_otp_enabled.find_each do |user|
+      # Gets the new value on already-updated users
+      # Falls back to legacy value on not-yet-migrated users
+      otp_secret = user.otp_secret
```

Comment:
How does it fall back to the legacy value?

Reply:
It's built-in behavior in the gem -- tries to read from column first, and then falls back to reading from the legacy calculated value when empty.

Reply:
But `MigrationUser` isn't using the gem, is it?

Reply:
One issue is that it makes the migration script very brittle and it will likely break with future code/schema updates.

I think we can reduce the changes of that happening while keeping the maintenance burden low by keeping a `MigrationUser` stub but making it use Devise.

If the maintenance burden does become too much in the future, I guess we can make an exception for this migration and have an explicit migration breakpoint, or offer alternatives (e.g. rake task) to admins.

In any case, I think a migration test is needed so that we catch this issue and dropping the migration path is a conscious choice that we can document, instead of something breaking without us being aware.

Reply:
I'll try this if you don't mind.

Reply:
I've pushed a test and fixed the following issues with the post-deployment migration:
- over-reliance on `User` model causing failures in some upgrade paths
- it could overwrite post-update OTP secrets with pre-update OTP secrets if users reset OTP in the meantime